src/doc/trpl/dining-philosophers.md

```diff
@@ -0,0 +1,942 @@
+all the philosophers, calling it:
+
+```rust
+struct Philosopher {
+    name: String,
+}   
+
+impl Philosopher { 
+    fn new(name: &str) -> Philosopher {
+        Philosopher {
+            name: name.to_string(),
+        }
+    }
+    
+    fn done(&self) {
+        println!("{} is done eating.", self.name);
+    }
+}
+
+fn main() {
+    let philosophers = vec![
+        Philosopher::new("Baruch Spinoza"),
+        Philosopher::new("Gilles Deleuze"),
+        Philosopher::new("Karl Marx"),
+        Philosopher::new("Friedrich Nietzsche"),
+        Philosopher::new("Michel Foucault"),
+    ];
+
+    for p in &philosophers {
+        p.done();
+    }
+}
+```
+
+Let’s look at `main()` first. Rather than have five individual variable
+bindings for our philosophers, we make a `Vec<T>` of them instead. `Vec<T>` is
+also called a ‘vector’, and it’s a growable array type. We then use a
+[`for`][for] loop to iterate through the vector, getting a reference to each
+philosopher in turn.
+
+[for]: for-loops.html
+
+In the body of the loop, we call `p.done()`, which is defined above:
+
+```rust,ignore
+fn done(&self) {
+    println!("{} is done eating.", self.name);
+}
+```
+
+In Rust, methods take an explicit `self` parameter. That’s why `done()` is a
+method, but `new` is an associated function: `new()` has no `self`. For our
+first version of `done()`, we just print out the name of the philosopher, and
+mention they’re done eating. Running this program should give you the following
+output:
+
+```text
+Baruch Spinoza is done eating.
+Gilles Deleuze is done eating.
+Karl Marx is done eating.
+Friedrich Nietzsche is done eating.
+Michel Foucault is done eating.
+```
+
+Easy enough, they’re all done! We haven’t actually implemented the real problem
+yet, though, so we’re not done yet!
+
+The next version is a bit larger. Here it is:
+
+```rust
+use std::sync::mpsc;
+use std::sync::mpsc::Sender;
+
+struct Philosopher {
+    name: String,
+    done: Sender<bool>
+}   
+
+impl Philosopher { 
+    fn new(name: &str, done: Sender<bool>) -> Philosopher {
+        Philosopher {
+            name: name.to_string(),
+            done: done,
+        }
+    }
+    
+    fn done(&self) {
+        println!("{} is done eating.", self.name);
+
+        self.done.send(true).ok().expect("Couldn't finish eating");
+    }
+}
+
+fn main() {
+    let (done_tx, done_rx) = mpsc::channel();
+
+    let philosophers = vec![
+        Philosopher::new("Baruch Spinoza", done_tx.clone()),
+        Philosopher::new("Gilles Deleuze", done_tx.clone()),
+        Philosopher::new("Karl Marx", done_tx.clone()),
+        Philosopher::new("Friedrich Nietzsche", done_tx.clone()),
+        Philosopher::new("Michel Foucault", done_tx.clone()),
+    ];
+
+    for p in &philosophers {
+        p.done();
+    }
+
+    for _ in 0..5 {
+        done_rx.recv().unwrap();
+    }
+}
+```
+
+Whew! Lots there. Let’s go through it:
+
+```rust,ignore
+use std::sync::mpsc;
+use std::sync::mpsc::Sender;
```

Comment:
Conventionally I'd personally rewrite these two imports as `use std::sync::mpsc::{self, Sender};`, but I'm also not sure if you wanted to specifically avoid `self` here or not.